patroni/ctl.py

```diff
@@ -1413,3 +1415,278 @@ def format_pg_version(version: int) -> str:
+    - ``postgresql.authentication``:
+        - superuser and replication users defined (if possible, usernames are set from the respective Patroni ENV vars,
+          otherwise the default 'postgres' and 'replicator' values are used).
+          If not a sample config, either DSN or PG ENV vars are used to define superuser authentication parameters.
+        - rewind user is defined for a sample config if PG version can be defined and PG version is 11+
+          (if possible, username is set from the respective Patroni ENV var)
+    - ``bootsrtap.dcs.postgresql.use_pg_rewind set to True if PG version is 11+
+    - ``postgresql.pg_hba`` defaults or the lines gathered from the source instance's hba_file
+    - ``postgresql.pg_ident`` the lines gathered from the source instance's ident_file
+
+    :param scope: Scope parameter value to write into the configuration.
+    :param file: Full path to the configuration file to be created (/tmp/patroni.yml by default).
+    :param sample: Optional flag. If set, no source instance will be used - generate config with some sane defaults.
+    :param dsn: Optional DSN string for the local instance to get GUC values from.
+    :param bin_dir: Optional path to Postgres binaries.
+    """
+    from patroni.config import Config
+    from patroni.validator import get_major_version
+
+    no_value_msg = '#FIXME'
+    pg_version = None
+
+    dynamic_config = Config.get_default_config()
+    dynamic_config['postgresql']['parameters'] = dict(dynamic_config['postgresql']['parameters'])
+    config: Dict[str, Any] = {
+        'scope': scope or no_value_msg,
+        'name': socket.gethostname(),
+        'bootstrap': {
+            'dcs': dynamic_config
+        },
+        'postgresql': {
+            'parameters': None,
+            'connect_address': no_value_msg,
+            'listen': no_value_msg,
+        },
+    }
+
+    if not sample:
+        enrich_config_from_running_instance(config, no_value_msg, dsn)
+
+    # some work around bin_dir definition
+    if sample and bin_dir:
+        config['postgresql']['bin_dir'] = bin_dir
+    elif not sample:
+        config['postgresql']['bin_dir'] = get_bin_dir_from_running_instance(config['postgresql']['data_dir'])
+
+    if 'bin_dir' in config['postgresql']:
+        # obtain version from the binary
+        try:
+            pg_version = postgres_major_version_to_int(get_major_version(config['postgresql']['bin_dir'] or None))
+        except PatroniException as e:
+            raise PatroniCtlException(str(e))
+    else:
+        config['postgresql']['bin_dir'] = ''
+
+    # generate sample config
+    if sample:
+        # some sane defaults or values set via Patroni env vars
+        replicator = os.getenv('PATRONI_REPLICATION_USERNAME', 'replicator')
+        config['postgresql']['authentication'] = {
+            'superuser': {
+                'username': os.getenv('PATRONI_SUPERUSER_USERNAME', 'postgres'),
```

Comment:
🤦‍♀️  `os.getenv` is useless with Patroni env vars here, as they all are pop'ed out of environ [here](https://github.com/zalando/patroni/blob/master/patroni/ctl.py#L195)